sys/kern/lockdep.c

```diff
@@ -0,0 +1,278 @@
+  bfs_mark_visited(src);
+
+  while ((link = bfs_next_link(link)) || (link = bfs_q_dequeue(queue))) {
+    if (bfs_is_visited(link->to))
+      continue;
+    bfs_mark_visited(link->to);
+
+    if (link->to == target)
+      return true;
+
+    if ((child = SIMPLEQ_FIRST(&link->to->locked_after))) {
+      if (!bfs_q_enqueue(queue, child))
+        panic("lockdep: bfs queue overflow");
+    }
+  }
+
+  return false;
+}
+
+static lock_class_t *look_up_lock_class(lock_class_mapping_t *lock) {
+  lock_class_key_t *key = lock->key;
+  lock_class_t *class;
+
+  SIMPLEQ_FOREACH(class, CLASS_HASH_CHAIN(key), hash_entry) {
+    if (class->key == key)
+      return class;
+  }
+
+  return NULL;
+}
+
+static lock_class_t *alloc_class(void) {
+  if (class_cnt >= MAX_CLASSES)
+    panic("lockdep: no more classes");
+
+  return &lock_classes[class_cnt++];
+}
+
+static lock_class_t *get_or_create_class(lock_class_mapping_t *lock) {
+  lock_class_t *class;
+
+  /* If the lock doesn't have a key then it is statically allocated. In this
+   * case use its address as the key. */
+  if (lock->key == NULL)
+    lock->key = (void *)lock;
+
+  class = look_up_lock_class(lock);
+  if (!class) {
+    class = alloc_class();
+    class->key = lock->key;
+    class->name = lock->name;
+    SIMPLEQ_INIT(&class->locked_after);
+    SIMPLEQ_INSERT_HEAD(CLASS_HASH_CHAIN(lock->key), class, hash_entry);
+  }
+  lock->lock_class = class;
+
+  return class;
+}
+
+static lock_class_link_t *alloc_link(void) {
+  if (link_cnt >= MAX_LINKS)
+    panic("lockdep: no more links");
+
+  return &lock_links[link_cnt++];
+}
+
+static void add_prev_link(void) {
+  lock_class_t *hprev, *hcur;
+  lock_class_link_t *link;
+  thread_t *thread = thread_self();
+  int depth = thread->td_lock_depth;
+
+  if (depth < 2)
+    return;
+
+  hcur = thread->td_held_locks[depth - 1];
+  hprev = thread->td_held_locks[depth - 2];
+
+  if (hcur == hprev)
+    return;
+
+  SIMPLEQ_FOREACH(link, &(hprev->locked_after), entry) {
+    if (hcur == link->to)
+      return;
+  }
+
+  link = alloc_link();
+  link->to = hcur;
+
+  SIMPLEQ_INSERT_HEAD(&(hprev->locked_after), link, entry);
+
+  if (check_path(hcur, hprev)) {
+    panic("lockdep: cycle between locks %s and %s", hprev->name, hcur->name);
+  }
+}
+
+void lockdep_init(void) {
+  for (int i = 0; i < CLASSHASH_SIZE; i++) {
+    SIMPLEQ_INIT(&lock_hashtbl[i]);
+  }
+}
+
+void lockdep_acquire(lock_class_mapping_t *lock) {
+  lock_class_t *class;
+  thread_t *thread = thread_self();
+
+  lockdep_lock();
+
+  if (thread->td_lock_depth >= LOCKDEP_MAX_HELD_LOCKS)
+    panic("lockdep: max lock depth reached");
+
+  if (!(class = lock->lock_class))
+    class = get_or_create_class(lock);
+
+  thread->td_held_locks[thread->td_lock_depth++] = class;
+
+  add_prev_link();
+
+  lockdep_unlock();
+}
+
+void lockdep_release(lock_class_mapping_t *lock) {
+  lock_class_t *class = lock->lock_class;
+  thread_t *thread = thread_self();
+  assert(class);
+
+  lockdep_lock();
```

Comment:
Do we even need to acquire the lock here? We're only accessing data that's private to the calling thread.

Reply:
Yeah, we only access fields from the thread and no one else can modify it beside that calling one.

Reply:
Actually, if we don't acquire the lock then we can receive an interrupt, but interrupt handlers can only use spinlocks, which are not (currently) checked by lockdep, so I don't think there is an issue, but there should be a comment mentioning it.